source/tutorials/callpackage.md

```diff
@@ -0,0 +1,283 @@
+---
+date: 2022-09-08
+authors:
+  - Norbert Melzer
+  - Valentin Gagarin
+  - Matthias Meschede
+myst:
+  html_meta:
+    "keywords": "tutorial, callPackage, override, package, customise, parameters, nix, nixpkgs"
+---
+
+(callpackage-tutorial)=
+# Package parameters and overrides with `callPackage`
+
+Nix ships with a a special-purpose programming language for creating packages and configurations: the Nix language.
+It is used to build the Nix package collection, known as {term}`Nixpkgs`.
+
+Being purely functional, the Nix language allows declaring custom functions to abstract over common patterns.
+One of the most prominent patterns in Nixpkgs is parametrisation of package recipes.
+
+## Overview
+
+Nixpkgs is sizeable software project on it's own, with coding conventions and idioms that have emerged over the years.
+It has [established a convention](https://github.com/NixOS/nixpkgs/pull/9869) of composing parameterised packages with automatic settings through a function named [`callPackage`](https://github.com/NixOS/nixpkgs/commit/fd268b4852d39c18e604c584dd49a611dc795a9b).
+This tutorial shows how to use it and why it's beneficial.
+
+### What will you learn?
+
+- Using `callPackage` to invoke package recipes that follow Nixpkgs conventions
+- Overriding package parameters
+- Creating interdependent package sets
+
+### What do you need?
+
+- Familiarity with the [Nix language](reading-nix-language)
+- First experience with [packaging existing software](packaging-tutorial)
+
+### How long does it take?
+
+- 45 minutes
+
+## Automatic function calls
+
+Create a new file `hello.nix`, which could be a typical package recipe as found in Nixpkgs:
+A function that takes an attribute set, with attributes corresponding to derivations in the top-level package set, and returns a derivation.
+
+```{code-block} nix
+:caption: hello.nix
+{ writeShellScriptBin }:
+writeShellScriptBin "hello" ''
+  echo "Hello, world!"
+''
+```
+
+:::{dropdown} Detailed explanation
+`hello.nix` declares a function which takes as argument an attribute set with one element `writeShellScriptBin`.
+[`writeShellScriptBin`](https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-writeShellScriptBin) is a function that happens to exist in Nixpkgs, a [build helper](https://nixos.org/manual/nixpkgs/unstable/#part-builders) that returns a derivation.
+The derivation output in this case is an executable shell script with the contents `echo "hello world"` named `"hello"`.
```

Comment:
```suggestion
The derivation output in this case contains an executable shell script in `$out/bin/hello` that prints "Hello world" when run.
```

The derivation output would be the entire `$out` directory, not just `$out/bin/hello`.

And it doesn't just contain that string, it also contains a shebang before that.